p4-16/psa/PSA.mdk

```diff
@@ -590,6 +590,92 @@ stream (copying or redirecting).
+actions associated with them.  They may only be used to determine
+whether a particular value is in the set, returning a Boolean value.
+That Boolean value can then be used in a `select` statement to control
+parsing (see examples below).
+
+```
+[INCLUDE=psa.p4:ValueSet_extern]
+```
+
+The control plane API excerpt above is intended to be added as part of
+the P4Runtime API[^P4RuntimeAPI].
+
+[^P4RuntimeAPI]: The P4Runtime API, defined as a Google Protocol
+    Buffer `.proto` file, can be found at
+    <https://github.com/p4lang/PI/blob/master/proto/p4/p4runtime.proto>
+
+The control plane API for a `ValueSet` is similar to that of a table,
+except only match fields may be specified, with no actions.  This
+includes API calls that specify ternary or range matching, although
+for `ValueSet`s these do not require specifying any priority values,
+since the only result of a `ValueSet` `is_member` call is "in the set"
+or "not in the set".
+
+If a PSA target can do so, it should implement control plane API calls
+involving ternary or range matching using ternary or range matching
+capabilities in the target, consuming the minimal table entries
+possible.
+
+However, a PSA target is allowed to implement such control plane API
+calls by "expanding" them into as many exact match entries as needed
+to have the same behavior.  For example, a control plane API call
+adding all values in the range 5 through 8 may be implemented as
+adding the four separate exact match values 5, 6, 7, and 8.
+
+The parser definition below shows an example that uses two `ValueSet`
+instances called `tpid_types` and `trill_types`.
+
+```
+[INCLUDE=examples/psa-example-value-sets.p4:ValueSet_Example_1]
+```
+
+The second example (below) has the same parsing behavior as the
+example above, but combines the two parse states
+`dispatch_tpid_value_set` and `dispatch_trill_value_set` into one.
+
+```
+[INCLUDE=examples/psa-example-value-sets2.p4:ValueSet_Example_2]
+```
+
+The third example (below) demonstrates one way to have a `ValueSet`
+that matches on multiple fields, by making the type `D` a `struct`
+containing multiple bit vectors.
+
+```
+[INCLUDE=examples/psa-example-value-sets3.p4:ValueSet_Example_3]
+```
+
+TBD: Should PSA allow code in the style of the second example, where
+multiple boolean variables can be in the `select` expression?  I can't
+think of a strong reason why not.
+
+TBD: Should PSA allow programs to use the value `false` in a case of a
+`select` statement that is matching on a variable returned from an
+`is_member` method call?
+
+TBD: Should it be allowed to perform more complex calculations on the
+return value of `is_member` before using it in a `select` expression?
+Eventually I am sure some PSA implementations would (reasonably)
+reject such things as too complex.  Here is an example that might not
+be too complex, but you can easily imagine similar code that is more
+complex:
+```
```

Comment:
@vgurevich, I seem to recall that we also discussed this in the language and that assignment was one of the push backs. So, indeed, we should only allow `is_member` in a `select` expression. We can go to the "extreme" of rewriting Example 2 and inline the calls, so that we make it clearer that is the only supported use.